#!/bin/bash
# ~~~ Indexing the Reference ~~~ #
bowtie2-build data/sequences.fna alignment_based/reference_index

# ~~~ Aligning Synthetic Reads ~~~ #
# Directory containing the synthetic reads generated by Polyester
READ_DIR="polyester_reads"

mkdir -p alignment_based

# Declare conditions, error levels, and replicates
conditions=("control" "treatment")
error_levels=("low" "medium" "high")
replicates=("01" "02" "03")

# Loop through conditions, error levels, and replicates
for condition in "${conditions[@]}"; do
    for error_level in "${error_levels[@]}"; do
        for rep in "${replicates[@]}"; do
            # Construct the input file names based on condition, error level, and replicate
            input_file_1="polyester_reads/${condition}_${error_level}/shuffled_sample_${rep}_1.fasta"
            input_file_2="polyester_reads/${condition}_${error_level}/shuffled_sample_${rep}_2.fasta"
            
            # Construct the output file name based on condition, error level, and replicate
            output_sam="alignment_based/${condition}_${error_level}_${rep}.sam"
            
            # Run Bowtie2 alignment
            bowtie2 -f -x alignment_based/reference_index -1 $input_file_1 -2 $input_file_2 -S $output_sam

            # Run featureCounts
            featureCounts -a data/annotation.gtf -o alignment_based/${condition}_${error_level}_${rep}_counts.txt -t exon -g gene_id $output_sam
        done
    done
done